### ISSUE TITLE
Using DenseReparameterization inside symbolic loops

### ISSUE THREAD

**Issue by danijar** [tensorflow/probability#114]
Currently, the problem is that the divergence is only built once, during the first `layer.call()`. This means if the layer was already used in one control flow context and is then applied again in another, the `layer.losses` will still contain the divergence tensor from the previous control flow context, which cannot be accessed from inside the new one. 

My current workaround is to set `layer._built_kernel_divergence = False` before each invocation of the layer, and then add `layer.losses[-1]` to the loss function, which is the newly created divergence tensor.

**Comment by danijar** [comment]
By the way, `tf.layers` work in this scenario. So if the goal is to provide near drop-in replacements for `tf.layers`, this would be useful to fix. It should mainly be writing a few tests with probabilistic layers inside one or two control flow constructs and moving the workaround I described above into the implementation, unless there are good reasons it doesn't work this way right now?
